src/corecel/math/RegulaFalsiRootFinder.hh

```diff
@@ -0,0 +1,113 @@
+ *
+ * Perform a Regula Falsi Iteration given a root function \em func and
+ * toolerance \em tol .
+ *
+ * Using a \em left and \em right bound a Regula Falsi approximates the \em
+ * root as: \f[ root = (left * func(right) - right * func(left)) / (func(right)
+ * - func(left)) \f]
+ *
+ * Then value of \em func at the root is calculated compared to values of
+ * \em func at the bounds. The bound which has the same sign \em func(root)
+ */
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <cmath>
+#include <type_traits>
+
+#include "celeritas_config.h"
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/math/Algorithms.hh"
+
+#include "detail/MathImpl.hh"
+
+namespace celeritas
+{
+template<class F>
+class RegulaFalsi
+{
+  public:
+    // Contructpr of Regula Falsi
+    inline CELER_FUNCTION RegulaFalsi(F&& func, real_type tol);
+
+    // Solve for a root between two points
+    real_type operator()(real_type left, real_type right);
+
+  private:
+    F func_;
+    real_type tol_;
+};
+
+//---------------------------------------------------------------------------//
+// TEMPLATE DEDUCTION
+//---------------------------------------------------------------------------//
+
+template<class F, class... Args>
+RegulaFalsi(F&&, Args...) -> RegulaFalsi<F>;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct from function.
+ */
+template<class F>
+CELER_FUNCTION RegulaFalsi<F>::RegulaFalsi(F&& func, real_type tol)
+    : func_{celeritas::forward<F>(func)}, tol_{tol}
+{
+    CELER_EXPECT(tol > 0);
+}
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+
+//---------------------------------------------------------------------------//
+/*!
+ * Perform Regula Falsi in defined bounds \em left and \em right with
+ * parameters defined in \em params.
+ */
+template<class F>
+CELER_FUNCTION real_type RegulaFalsi<F>::operator()(real_type left,
+                                                    real_type right)
+{
+    // Initialize Iteration parameters
+    real_type f_left = func_(left);
+    real_type f_right = func_(right);
+    real_type f_root = 1;
+    real_type root = 0;
+    int i = 0;
+
+    // Iterate on root
+    do
+    {
+        // Calcuate root
+        root = (left * f_right - right * f_left) / (f_right - f_left);
+
+        // Root function value of root
+        f_root = func_(root);
+
+        // Update the bound which produces the same sign as the root
+        if (signum(f_left) == signum(f_root))
+        {
+            left = root;
+            f_left = f_root;
+        }
+        else
+        {
+            right = root;
+            f_right = f_root;
+        }
+        i++;
+    } while (std::fabs(f_root) > tol_ && i <= 100);
```

Comment:
Instead of a magic value of 100, define a `static constexpr inline int max_iters_ = 100;` in the class. In addition, a trick I learned to make checking the termination criterion slightly more efficient in some circumstances, is to initialize them with the maximum value and then decrement until the value is negative:

```suggestion
    } while (std::fabs(f_root) > tol_ && --remaining_iters > 0);
```

we could also consider changing the loop criterion to an assertion, otherwise you're silently returning an unconverged value....

Reply:
Do you mean like putting an ensure statement after the loop to make sure the convergence criteria is met?